File: tokenize_xml.py
from transformers import AutoTokenizer

def tokenize_xml(xml_file, model_name='sentence-transformers/all-MiniLM-L6-v2'):
    """
    Tokenizes an XML file using a HuggingFace sentence-transformers model tokenizer.
    Shows detailed information about tokens, splits, and token count.

    Args:
        xml_file (str): Path to the XML file to be tokenized.
        model_name (str): Name of the sentence-transformers model to use.
    """
    # Load the tokenizer
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    
    # Read the XML file
    with open(xml_file, 'r', encoding='utf-8') as file:
        xml_content = file.read()
    
    print("=" * 80)
    print(f"MODEL: {model_name}")
    print("=" * 80)
    
    # Tokenize without truncation to see full token count
    tokens_full = tokenizer(xml_content, truncation=False, add_special_tokens=True)
    
    # Get token IDs
    token_ids = tokens_full['input_ids']
    
    # Convert token IDs back to tokens to see how text is split
    tokens = tokenizer.convert_ids_to_tokens(token_ids)
    
    # Get the original text for each token
    print("\n" + "=" * 80)
    print("TOKENIZATION DETAILS")
    print("=" * 80)
    
    print(f"\n📊 TOTAL TOKEN COUNT: {len(tokens)}")
    print(f"📊 Token IDs count: {len(token_ids)}")
    print(f"📊 Model max length: {tokenizer.model_max_length}")
    
    # Check if truncation would occur
    if len(tokens) > tokenizer.model_max_length:
        print(f"\n⚠️  WARNING: Content exceeds model max length!")
        print(f"   Tokens will be truncated from {len(tokens)} to {tokenizer.model_max_length}")
    
    print("\n" + "-" * 80)
    print("TOKEN LIST (showing how text is split into subwords)")
    print("-" * 80)
    
    # Display tokens with their IDs in a formatted way
    print(f"\n{'Index':<8} {'Token ID':<12} {'Token':<30}")
    print("-" * 50)
    
    for i, (token_id, token) in enumerate(zip(token_ids, tokens)):
        # Highlight special tokens
        if token in ['[CLS]', '[SEP]', '[PAD]', '[UNK]', '<s>', '</s>', '<pad>']:
            print(f"{i:<8} {token_id:<12} {token:<30} [SPECIAL]")
        else:
            print(f"{i:<8} {token_id:<12} {token:<30}")
    
    print("\n" + "=" * 80)
    print("TOKEN ANALYSIS")
    print("=" * 80)
    
    # Analyze subword splitting
    subword_tokens = [t for t in tokens if t.startswith('##') or t.startswith('▁')]
    regular_tokens = [t for t in tokens if not t.startswith('##') and not t.startswith('▁') 
                      and t not in ['[CLS]', '[SEP]', '[PAD]', '[UNK]', '<s>', '</s>', '<pad>']]
    special_tokens = [t for t in tokens if t in ['[CLS]', '[SEP]', '[PAD]', '[UNK]', '<s>', '</s>', '<pad>']]
    
    print(f"\n📈 Token Statistics:")
    print(f"   - Total tokens: {len(tokens)}")
    print(f"   - Regular tokens: {len(regular_tokens)}")
    print(f"   - Subword tokens (## prefix): {len(subword_tokens)}")
    print(f"   - Special tokens: {len(special_tokens)}")
    
    # Show unique tokens
    unique_tokens = set(tokens)
    print(f"   - Unique tokens: {len(unique_tokens)}")
    
    # Decode back to text to show reconstruction
    print("\n" + "-" * 80)
    print("DECODED TEXT (reconstructed from tokens)")
    print("-" * 80)
    decoded_text = tokenizer.decode(token_ids, skip_special_tokens=True)
    print(f"\n{decoded_text[:]}...")  # Show chars
    
    return {
        'tokens': tokens,
        'token_ids': token_ids,
        'token_count': len(tokens),
        'unique_token_count': len(unique_tokens),
        'model_max_length': tokenizer.model_max_length
    }


if __name__ == "__main__":
    result = tokenize_xml('example_xml.txt')
    print("\n" + "=" * 80)
    print("SUMMARY")
    print("=" * 80)
    print(f"Total tokens: {result['token_count']}")
    print(f"Unique tokens: {result['unique_token_count']}")
    print(f"Model max length: {result['model_max_length']}")

File: tokenize_xml.ts
import * as fs from 'fs';
import { AutoTokenizer } from '@xenova/transformers';

interface TokenizationResult {
  tokens: string[];
  token_ids: number[];
  token_count: number;
  unique_token_count: number;
  model_max_length: number;
}

async function tokenizeXml(
  xmlFile: string,
  modelName: string = 'sentence-transformers/all-MiniLM-L6-v2'
): Promise<TokenizationResult> {
  /**
   * Tokenizes an XML file using a HuggingFace sentence-transformers model tokenizer.
   * Shows detailed information about tokens, splits, and token count.
   */
  
  // Load the tokenizer
  console.log(`Loading tokenizer for: ${modelName}`);
  const tokenizer = await AutoTokenizer.from_pretrained(modelName);
  
  // Read the XML file
  const xmlContent = fs.readFileSync(xmlFile, 'utf-8');
  
  console.log('='.repeat(80));
  console.log(`MODEL: ${modelName}`);
  console.log('='.repeat(80));
  
  // Tokenize the content
  const encoded = await tokenizer(xmlContent);
  const tokenIds = Array.from(encoded.input_ids.data) as number[];
  
  // Decode individual tokens
  const tokens: string[] = [];
  for (const tokenId of tokenIds) {
    const token = tokenizer.decode([tokenId], { skip_special_tokens: false });
    tokens.push(token);
  }
  
  // Get the model's max length
  const modelMaxLength = tokenizer.model_max_length || 512;
  
  console.log('\n' + '='.repeat(80));
  console.log('TOKENIZATION DETAILS');
  console.log('='.repeat(80));
  
  console.log(`\n📊 TOTAL TOKEN COUNT: ${tokens.length}`);
  console.log(`📊 Token IDs count: ${tokenIds.length}`);
  console.log(`📊 Model max length: ${modelMaxLength}`);
  
  // Check if truncation would occur
  if (tokens.length > modelMaxLength) {
    console.log(`\n⚠️  WARNING: Content exceeds model max length!`);
    console.log(`   Tokens will be truncated from ${tokens.length} to ${modelMaxLength}`);
  }
  
  console.log('\n' + '-'.repeat(80));
  console.log('TOKEN LIST (showing how text is split into subwords)');
  console.log('-'.repeat(80));
  
  // Display tokens with their IDs in a formatted way
  console.log(`\n${'Index'.padEnd(8)} ${'Token ID'.padEnd(12)} ${'Token'.padEnd(30)}`);
  console.log('-'.repeat(50));
  
  const specialTokens = ['[CLS]', '[SEP]', '[PAD]', '[UNK]', '<s>', '</s>', '<pad>'];
  
  for (let i = 0; i < tokenIds.length; i++) {
    const tokenId = tokenIds[i];
    const token = tokens[i];
    
    // Highlight special tokens
    if (specialTokens.includes(token)) {
      console.log(`${String(i).padEnd(8)} ${String(tokenId).padEnd(12)} ${token.padEnd(30)} [SPECIAL]`);
    } else {
      console.log(`${String(i).padEnd(8)} ${String(tokenId).padEnd(12)} ${token.padEnd(30)}`);
    }
  }
  
  console.log('\n' + '='.repeat(80));
  console.log('TOKEN ANALYSIS');
  console.log('='.repeat(80));
  
  // Analyze subword splitting
  const subwordTokens = tokens.filter(t => t.startsWith('##') || t.startsWith('▁'));
  const regularTokens = tokens.filter(
    t => !t.startsWith('##') && !t.startsWith('▁') && !specialTokens.includes(t)
  );
  const specialTokensList = tokens.filter(t => specialTokens.includes(t));
  
  console.log(`\n📈 Token Statistics:`);
  console.log(`   - Total tokens: ${tokens.length}`);
  console.log(`   - Regular tokens: ${regularTokens.length}`);
  console.log(`   - Subword tokens (## or ▁ prefix): ${subwordTokens.length}`);
  console.log(`   - Special tokens: ${specialTokensList.length}`);
  
  // Show unique tokens
  const uniqueTokens = new Set(tokens);
  console.log(`   - Unique tokens: ${uniqueTokens.size}`);
  
  // Decode back to text to show reconstruction
  console.log('\n' + '-'.repeat(80));
  console.log('DECODED TEXT (reconstructed from tokens)');
  console.log('-'.repeat(80));
  
  const decodedText = tokenizer.decode(tokenIds, { skip_special_tokens: true });
  console.log(`\n${decodedText}...`);
  
  return {
    tokens,
    token_ids: tokenIds,
    token_count: tokens.length,
    unique_token_count: uniqueTokens.size,
    model_max_length: modelMaxLength
  };
}

// Main execution - check if this file is being run directly
const isMainModule = import.meta.url === `file://${process.argv[1]}`;

if (isMainModule) {
  (async () => {
    const result = await tokenizeXml('example_xml.txt');
    console.log('\n' + '='.repeat(80));
    console.log('SUMMARY');
    console.log('='.repeat(80));
    console.log(`Total tokens: ${result.token_count}`);
    console.log(`Unique tokens: ${result.unique_token_count}`);
    console.log(`Model max length: ${result.model_max_length}`);
  })().catch(console.error);
}

export { tokenizeXml, TokenizationResult };
